dotnet/src/Experimental/Process.LocalRuntime/LocalMap.cs

```diff
@@ -0,0 +1,135 @@
+        this._map = map;
+        this._logger = this._kernel.LoggerFactory?.CreateLogger(this._map.State.Name) ?? new NullLogger<LocalStep>();
+        this._mapEvents = [.. map.Edges.Keys.Select(key => key.Split(ProcessConstants.EventIdSeparator).Last())];
+    }
+
+    /// <inheritdoc/>
+    internal override async Task HandleMessageAsync(ProcessMessage message)
+    {
+        // Initialize the current operation
+        (IEnumerable inputValues, KernelProcess mapOperation, string startEventId) = this._map.Initialize(message, this._logger);
+
+        // Prepare state for map execution
+        int index = 0;
+        List<(Task Task, LocalKernelProcessContext ProcessContext, MapOperationContext Context)> mapOperations = [];
+        ConcurrentDictionary<string, Type> capturedEvents = [];
+        try
+        {
+            // Execute the map operation for each value
+            foreach (var value in inputValues)
+            {
+                ++index;
+
+                KernelProcess process = mapOperation.CloneProcess(this._logger);
+                MapOperationContext context = new(this._mapEvents, capturedEvents);
+#pragma warning disable CA2000 // Dispose objects before losing scope
+                LocalKernelProcessContext processContext = new(process, this._kernel, context.Filter);
+                Task processTask =
+                    processContext.StartWithEventAsync(
+                        new KernelProcessEvent
+                        {
+                            Id = startEventId,
+                            Data = value
+                        });
+#pragma warning restore CA2000 // Dispose objects before losing scope
+
+                mapOperations.Add((processTask, processContext, context));
+            }
+
+            // Wait for all the map operations to complete
+            await Task.WhenAll(mapOperations.Select(p => p.Task)).ConfigureAwait(false);
+
+            // Correlate the operation results to emit as the map result
+            Dictionary<string, Array> resultMap = [];
+            for (index = 0; index < mapOperations.Count; ++index)
+            {
+                foreach (KeyValuePair<string, Type> capturedEvent in capturedEvents)
+                {
+                    string eventName = capturedEvent.Key;
+                    Type resultType = capturedEvent.Value;
+
+                    mapOperations[index].Context.Results.TryGetValue(eventName, out object? result);
+                    if (!resultMap.TryGetValue(eventName, out Array? results))
+                    {
+                        results = Array.CreateInstance(resultType, mapOperations.Count);
+                        resultMap[eventName] = results;
+                    }
+
+                    results.SetValue(result, index);
+                }
+            }
+
+            // Emit map results
+            foreach (string eventName in capturedEvents.Keys)
+            {
+                Array eventResult = resultMap[eventName];
```

Comment:
`eventResults` would be a better name

Reply:
I hear where you are coming from.  The cardinality aspects at play here a bit complex.  Each map-step "result" is always an enumeration and a map operation may emit multiple events or "results"...which would be multiple enumerations.

This is a single enumeration or "result" that is associated with a single event.  The for loop is emitting multiple "results" (as commented).